logaggregator/rfc5424/rfc5424.go

```diff
@@ -0,0 +1,180 @@
+package rfc5424
+
+import (
+	"bytes"
+	"strconv"
+	"time"
+)
+
+type Message struct {
+	Facility       int
+	Severity       int
+	Version        int
+	Timestamp      time.Time
+	Hostname       string
+	AppName        string
+	ProcID         string
+	MsgID          string
+	StructuredData string
+	Msg            string
+}
+
+func (m Message) PriVal() int {
+	return m.Facility*8 + m.Severity
+}
+
+func Parse(buf []byte) (*Message, error) {
+	cursor := 0
+	msg := &Message{}
+	if err := parseHeader(buf, &cursor, msg); err != nil {
+		return nil, err
+	}
+
+	if err := parseStructuredData(buf, &cursor, msg); err != nil {
+		return nil, err
+	}
+
+	if cursor < len(buf) {
+		msg.Msg = string(buf[cursor:])
+	}
+
+	return msg, nil
+}
+
+type ParseError struct {
+	Cursor  int
+	Message string
+}
+
+func (p *ParseError) Error() string {
+	return p.Message
```

Comment:
Convention is to prefix the error string with the name of the package it came from:

```go
return "rfc5424: "+p.Message
```